// @orqastudio/types/dist/project.js
/** Default per-type colours for artifact link chips. */
export const DEFAULT_ARTIFACT_LINK_COLORS = {
    EPIC: "#3b82f6",
    TASK: "#06b6d4",
    RULE: "#a78bfa",
    AD: "#8b5cf6",
    IDEA: "#c084fc",
    IMPL: "#67e8f9",
    KNOW: "#2dd4bf",
    PILLAR: "#818cf8",
    RES: "#6366f1",
    MS: "#38bdf8",
    DOC: "#94a3b8",
    AGENT: "#f472b6",
};
/** Type guard: is this entry a group (has children)? */
export function isArtifactGroup(entry) {
    return "children" in entry;
}
// Import the platform config data file. This is the single source of truth
// for core artifact types and relationships. Plugins and project config
// extend this at runtime — the platform config is not special-cased in code.
import platformConfigData from "./platform/core.json" with { type: "json" };
/** The loaded platform configuration. */
export const PLATFORM_CONFIG = platformConfigData;
/**
 * Platform artifact type keys derived from core.json.
 * Backwards-compatible: same shape as the old hardcoded array.
 */
export const PLATFORM_ARTIFACT_TYPES = PLATFORM_CONFIG.artifactTypes.map(t => t.key);
/**
 * Platform relationships derived from core.json.
 * Backwards-compatible: same RelationshipType[] shape.
 */
export const PLATFORM_RELATIONSHIPS = PLATFORM_CONFIG.relationships;
/**
 * Semantic categories from core.json.
 * Used by checks to query relationship intent without hardcoding keys.
 */
export const PLATFORM_SEMANTICS = PLATFORM_CONFIG.semantics;
/**
 * Default platform navigation groups (principles, learning, discovery).
 * These are the builtin groups that every project starts with.
 * Plugins add their own groups/items via defaultNavigation.
 */
export const PLATFORM_NAVIGATION = [
    { key: "project", type: "builtin", icon: "layout-dashboard" },
    {
        key: "principles",
        type: "group",
        icon: "landmark",
        label: "Principles",
        children: [
            { key: "pillars", type: "builtin", icon: "columns-3" },
            { key: "vision", type: "builtin", icon: "eye" },
            { key: "personas", type: "builtin", icon: "users" },
            { key: "grounding", type: "builtin", icon: "anchor" },
        ],
    },
    {
        key: "discovery",
        type: "group",
        icon: "compass",
        label: "Discovery",
        children: [
            { key: "ideas", type: "builtin", icon: "lightbulb" },
        ],
    },
    {
        key: "learning",
        type: "group",
        icon: "brain",
        label: "Learning",
        children: [
            { key: "decisions", type: "builtin", icon: "scale" },
            { key: "rules", type: "builtin", icon: "shield" },
            { key: "lessons", type: "builtin", icon: "book-open" },
            { key: "knowledge", type: "builtin", icon: "zap" },
            { key: "agents", type: "builtin", icon: "bot" },
        ],
    },
    { key: "artifact-graph", type: "builtin", icon: "network" },
    { key: "settings", type: "builtin", icon: "settings" },
];
